Documentation/design-docs/standalone-gc-eventing.md

```diff
@@ -0,0 +1,212 @@
+    /* IN */ void* eventHandle,
+    /* IN */ void** eventData,
+    /* IN */ uint32_t eventDataCount,
+    /* IN */ const GUID* pActivityId = nullptr,
+    /* IN */ const GUID* pRelatedActivityId = nullptr
+);
+```
+
+The four arguments of this function are:
+1. `eventHandle` - An opaque handle to the EE's representation of the event being fired.
+2. `eventData` - An array of pointers that contains the payload of the event being fired.
+Each `GCEventData` is a tuple of a data pointer and size, so it can be used to describe arbitrary data.
+3. `eventDataCount` - The size of the `eventData` array.
+4. `pActivityId` and `pRelatedActivityId` - If provided, passes the activity ID and related activity ID
+verbatim to the underlying event implementation.
+
+The EE will use this information to determine 1) if this event truly is enabled and, 2) if it is,
+forwarding it to the apporopriate underlying event implementation, be it EventPipe, ETW, or LTTNG.
+The passed-in eventHandle will be a handle to an object from which the EE can get information about
+the version of the event and other useful metadata.
+
+## Getting Informed of Changes to Event State
+
+There are three mechanisms by which CoreCLR is able to log events: EventPipe, ETW, and LTTNG. When it
+comes to changing the state of events, EventPipe and ETW both allow users to attach callbacks that are
+invoked whenever events are enabled or disabled. For these two mechanisms, it is sufficient to use
+this callback mechanism to call `IGCHeap::{Enable/Disable}Events` from within such a callback in order
+to inform the GC of changes to tracing state.
+
+LTTNG does not have such a mechanism. In order to observe changes in the eventing state, LTTNG must be
+polled periodically. For this purpose, when tracing using LTTNG, a new thread will be spawned by the EE
+whose responsibility is to periodically poll LTTNG and invoke the GC whenever it detects changes in the
+eventing state.
+
+## Adding New Events
+
+It is useful for a standalone GC to be able to fire events that the EE was not previously aware of.
+For example, it is useful for a GC developer to add some event-based instrumentation to the GC, especially
+when testing new features and ensuring that they work as expected.
+
+While it is possible for some eventing implementations to receive events that are created at runtime, not all
+eventing implementations (particularly LTTNG) are not flexible enough for this. In order to accomodate new
+events, another method is added to `IGCToCLR`:
+
+```c++
+void IGCToCLR::FireCustomEvent(
+    /* IN */ const char* eventName,
+    /* IN */ uint64_t eventId,
+    /* IN */ void* payload,
+    /* IN */ size_t payloadSize
+);
+```
+
+This callback fires a "custom" event that the EE is not previously aware of. The four arguments of this
+function are:
+
+1. `eventName` - The name of this custom event.
+2. `eventId` - A unique integer ID of this custom event that consumers can use to identify this event.
```

Comment:
You probably just need one of eventName or eventId.  I'd recommend name since collisions will be much less common.